<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Kim Dill-McFarland">
<meta name="dcterms.date" content="2023-12-04">

<title>Bulk RNAseq differential expression analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="2.Bulk_RNAseq_deg_analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/quarto.js"></script>
<script src="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="2.Bulk_RNAseq_deg_analysis_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="2.Bulk_RNAseq_deg_analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="2.Bulk_RNAseq_deg_analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="2.Bulk_RNAseq_deg_analysis_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


<link rel="stylesheet" href="styles.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#overview" id="toc-overview" class="nav-link active" data-scroll-target="#overview">Overview</a></li>
  <li><a href="#prior-to-the-tutorial" id="toc-prior-to-the-tutorial" class="nav-link" data-scroll-target="#prior-to-the-tutorial">Prior to the tutorial</a></li>
  <li><a href="#r-project" id="toc-r-project" class="nav-link" data-scroll-target="#r-project">R project</a></li>
  <li><a href="#load-packages" id="toc-load-packages" class="nav-link" data-scroll-target="#load-packages">Load packages</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a>
  <ul class="collapse">
  <li><a href="#data-description" id="toc-data-description" class="nav-link" data-scroll-target="#data-description">Data description</a></li>
  <li><a href="#copy-data" id="toc-copy-data" class="nav-link" data-scroll-target="#copy-data">Copy data</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  </ul></li>
  <li><a href="#introduction-to-linear-modeling" id="toc-introduction-to-linear-modeling" class="nav-link" data-scroll-target="#introduction-to-linear-modeling">Introduction to linear modeling</a></li>
  <li><a href="#modeling-in-limma" id="toc-modeling-in-limma" class="nav-link" data-scroll-target="#modeling-in-limma">Modeling in <code>limma</code></a>
  <ul class="collapse">
  <li><a href="#simple-linear-regression-in-limma" id="toc-simple-linear-regression-in-limma" class="nav-link" data-scroll-target="#simple-linear-regression-in-limma">Simple linear regression in <code>limma</code></a></li>
  <li><a href="#paired-sample-design-in-limma" id="toc-paired-sample-design-in-limma" class="nav-link" data-scroll-target="#paired-sample-design-in-limma">Paired sample design in <code>limma</code></a></li>
  </ul></li>
  <li><a href="#modeling-in-kimma" id="toc-modeling-in-kimma" class="nav-link" data-scroll-target="#modeling-in-kimma">Modeling in <code>kimma</code></a></li>
  <li><a href="#picking-a-best-fit-model" id="toc-picking-a-best-fit-model" class="nav-link" data-scroll-target="#picking-a-best-fit-model">Picking a best fit model</a></li>
  <li><a href="#significant-genes" id="toc-significant-genes" class="nav-link" data-scroll-target="#significant-genes">Significant genes</a></li>
  <li><a href="#navigation" id="toc-navigation" class="nav-link" data-scroll-target="#navigation"><em>Navigation</em></a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><em>Additional resources</em></a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Bulk RNAseq differential expression analysis</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Kim Dill-McFarland </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 4, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="overview" class="level2">
<h2 class="anchored" data-anchor-id="overview">Overview</h2>
<p>One common bulk RNAseq analysis is differential expression, which determines individual genes that significantly change by your variables of interest. Most commonly, you will see linear modelings with the R package <code>limma</code> and binomial models with <code>DeSeq2</code>. These and other packages are designed to run statistical models across 1000s of genes in an efficient manor. Here, we will explore <code>limma</code> as well as <code>kimma</code> which expands the RNAseq model framework to mixed effect and covariance models.</p>
</section>
<section id="prior-to-the-tutorial" class="level2">
<h2 class="anchored" data-anchor-id="prior-to-the-tutorial">Prior to the tutorial</h2>
<p>Please follow the setup instructions at <a href="https://fredhutch.github.io/seatrac-hackday-2023/1.rnaseq_tutorial/1.setup.html" class="uri">https://fredhutch.github.io/seatrac-hackday-2023/1.rnaseq_tutorial/1.setup.html</a>.</p>
</section>
<section id="r-project" class="level2">
<h2 class="anchored" data-anchor-id="r-project">R project</h2>
<p>First, create an R project for this tutorial. When prompted, don’t save your current <code>.RData</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>usethis<span class="sc">::</span><span class="fu">create_project</span>(<span class="at">path =</span> <span class="st">"1.rnaseq_tutorial"</span>, <span class="at">rstudio =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-packages" class="level2">
<h2 class="anchored" data-anchor-id="load-packages">Load packages</h2>
<p>Load the following packages.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Data manipulation and plotting</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Linear modeling</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kimma)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(limma)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">651</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<section id="data-description" class="level3">
<h3 class="anchored" data-anchor-id="data-description">Data description</h3>
<p>We will be using data from</p>
<blockquote class="blockquote">
<p>Darrah PA <em>et al</em>. Airway T-cells are a correlate of i.v. Bacille Calmette-Guerin-mediated protection against tuberculosis in rhesus macaques. Cell Host Microbe. 2023 Jun 14;31(6):962-977.e8. doi: <a href="https://doi.org/10.1016/j.chom.2023.05.006">10.1016/j.chom.2023.05.006</a>. PMID: 37267955; PMCID: PMC10355173.</p>
</blockquote>
<p>Specifically, we will be focusing on the T-cell transcriptional responses to <em>M. tuberculosis</em> (Mtb) challenge at week 25. For simplicity, we will ignore the vaccination groups, and you may explore them during your Hack time. These data are pseudo-bulk counts from the larger single-cell data set, which we will be exploring during the second part of this tutorial.</p>
<p>The raw counts data were cleaned and normalized following a standard pipeline. You can learn more at <a href="https://bigslu.github.io/tutorials/RNAseq/2.RNAseq_counts.to.voom.html" class="uri">https://bigslu.github.io/tutorials/RNAseq/2.RNAseq_counts.to.voom.html</a>. The code for cleaning of these data specifically is available at <a href="https://github.com/FredHutch/seatrac-hackday-2023/blob/main/1.rnaseq_tutorial/data_cleaning_bulk.R" class="uri">https://github.com/FredHutch/seatrac-hackday-2023/blob/main/1.rnaseq_tutorial/data_cleaning_bulk.R</a></p>
</section>
<section id="copy-data" class="level3">
<h3 class="anchored" data-anchor-id="copy-data">Copy data</h3>
<p>A copy of the entire tutorial is on the Hack Day server at <code>/home/seatrac-hackday-2023</code>.</p>
<p>Copy the tutorial data to a new <code>data/</code> directory in your R project.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"data"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>list.of.files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"/home/seatrac-hackday-2023/1.rnaseq_tutorial/data/"</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">file.copy</span>(list.of.files, <span class="st">"data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-data" class="level3">
<h3 class="anchored" data-anchor-id="load-data">Load data</h3>
<p>All counts, gene, and sample metadata are contained in a single object from the <code>limma</code> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"data/dat_tcell.RData"</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dat_tcell)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "genes"   "targets" "E"       "weights" "design" </code></pre>
</div>
</div>
<p>We access each data frame within this <code>Elist</code> using <code>$</code>. The normalized log2 CPM expression data are contained in <code>E</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dat_tcell<span class="sc">$</span>E[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        DDF16_WK25_STIMNO_T cell DDF16_WK25_STIMYES_T cell
A1BG                   0.1306417                 -3.419580
A2ML1                 -2.1912864                 -3.419580
A3GALT2               -2.1912864                 -1.834617</code></pre>
</div>
</div>
<p>Library and donor metadata are in <code>targets</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>dat_tcell<span class="sc">$</span>targets[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          group  lib.size norm.factors
DDF16_WK25_STIMNO_T cell      1 2283561.7    0.9200168
DDF16_WK25_STIMYES_T cell     1 5350150.0    1.0039990
DDF4E_WK25_STIMNO_T cell      1  877057.6    1.0069408
                                              libID  ptID week   mtb   cell
DDF16_WK25_STIMNO_T cell   DDF16_WK25_STIMNO_T cell DDF16 WK25 Media T cell
DDF16_WK25_STIMYES_T cell DDF16_WK25_STIMYES_T cell DDF16 WK25   Mtb T cell
DDF4E_WK25_STIMNO_T cell   DDF4E_WK25_STIMNO_T cell DDF4E WK25 Media T cell
                          sample.weights
DDF16_WK25_STIMNO_T cell       0.9555099
DDF16_WK25_STIMYES_T cell      0.7422811
DDF4E_WK25_STIMNO_T cell       0.4861639</code></pre>
</div>
</div>
<p>Gene metadata are in <code>genes</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>dat_tcell<span class="sc">$</span>genes[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           gene      gene_type     gene_stable_id
A1BG       A1BG protein_coding ENSMMUG00000012459
A2ML1     A2ML1 protein_coding ENSMMUG00000022680
A3GALT2 A3GALT2 protein_coding ENSMMUG00000003268</code></pre>
</div>
</div>
<p>Voom gene-level quality weights are in <code>weights</code>. These were calculated with <code>voomWithQualityWeights( )</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>example.voom<span class="sc">$</span>weights[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          [,1]      [,2]      [,3]
[1,] 0.7585106 0.8274172 1.5710160
[2,] 0.5288184 0.5555010 0.9553991
[3,] 0.7259811 0.7906774 1.5015908</code></pre>
</div>
</div>
<p>And finally, the null model used in voom normalization is found in <code>design</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>example.voom<span class="sc">$</span>design[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>lib1 lib2 lib3 
   1    1    1 </code></pre>
</div>
</div>
</section>
</section>
<section id="introduction-to-linear-modeling" class="level2">
<h2 class="anchored" data-anchor-id="introduction-to-linear-modeling">Introduction to linear modeling</h2>
<p>This tutorial assumes some familiarity with R and statistical modeling. You can find a quick intro relevant to RNA-seq data in <a href="https://bigslu.github.io/workshops/2023.01.30_RNAseq.i4TB/1_linear_models.html">Intro to linear modeling</a>.</p>
</section>
<section id="modeling-in-limma" class="level2">
<h2 class="anchored" data-anchor-id="modeling-in-limma">Modeling in <code>limma</code></h2>
<section id="simple-linear-regression-in-limma" class="level3">
<h3 class="anchored" data-anchor-id="simple-linear-regression-in-limma">Simple linear regression in <code>limma</code></h3>
<p><code>limma</code> take model inputs as a model matrix. This matrix encodes all the variables from the formula as 0 for N and 1 for Y. For example, here is the model matrix for the formula <code>~ mtb</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>mm_limma <span class="ot">&lt;-</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> mtb, <span class="at">data=</span>dat_tcell<span class="sc">$</span>targets)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(mm_limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                          (Intercept) mtbMtb
DDF16_WK25_STIMNO_T cell            1      0
DDF16_WK25_STIMYES_T cell           1      1
DDF4E_WK25_STIMNO_T cell            1      0
DDF4E_WK25_STIMYES_T cell           1      1
DDF4N_WK25_STIMNO_T cell            1      0
DDF4N_WK25_STIMYES_T cell           1      1</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p>Be careful with variable order! Note that we only see one level for each variable: Mtb for the mtb variable. This shows that the Mtb samples are being compared to the reference level (which is <code>mtb == "Media"</code>). The reference is determined alphabetically if the variable is a character and by level if the variable is a factor. So, if you want a different order than alphabetical, you need to format your variables as factors and set the appropriate order.</p>
</blockquote>
<p>Once we have a model matrix, we fit the model and estimate P-values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Fit model</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>fit_limma <span class="ot">&lt;-</span> <span class="fu">lmFit</span>(<span class="at">object =</span> dat_tcell<span class="sc">$</span>E, </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">design =</span> mm_limma,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">weights =</span> dat_tcell<span class="sc">$</span>weights)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Estimate significance</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>efit_limma <span class="ot">&lt;-</span> <span class="fu">eBayes</span>(<span class="at">fit =</span> fit_limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>These outputs contain a lot of information. We can pull out the most commonly used pieces with <code>topTable</code>. By default, this gives you the 10 most significant genes across the entire model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Extract results</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>fdr_limma <span class="ot">&lt;-</span> <span class="fu">topTable</span>(<span class="at">fit =</span> efit_limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Removing intercept from test coefficients</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fdr_limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           logFC  AveExpr        t      P.Value    adj.P.Val        B
PLXDC2  4.192477 6.786156 13.91593 1.083979e-15 1.218267e-11 25.45526
PLCB2   3.323085 6.618024 13.67945 1.790385e-15 1.218267e-11 24.99951
CLEC12A 3.679747 5.901780 12.90214 9.731346e-15 4.414463e-11 23.21073
JUND    1.568763 7.774542 12.66888 1.639179e-14 5.576896e-11 23.04095
ARHGDIB 1.829897 8.873965 12.09001 6.146497e-14 1.672954e-10 21.75189
LST1    2.556890 6.774452 11.68693 1.580100e-13 3.583929e-10 20.78367</code></pre>
</div>
</div>
<p>With some additional parameters, we can get gene results for individual variables. This is the same for this simple model but would differ if you had multiple variables of interest, covariates, interaction term, etc.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>fdr_limma_mtb <span class="ot">&lt;-</span> <span class="fu">topTable</span>(<span class="at">fit =</span> efit_limma, </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">coef =</span> <span class="st">"mtbMtb"</span>, <span class="at">number =</span> <span class="cn">Inf</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fdr_limma_mtb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           logFC  AveExpr        t      P.Value    adj.P.Val        B
PLXDC2  4.192477 6.786156 13.91593 1.083979e-15 1.218267e-11 25.45526
PLCB2   3.323085 6.618024 13.67945 1.790385e-15 1.218267e-11 24.99951
CLEC12A 3.679747 5.901780 12.90214 9.731346e-15 4.414463e-11 23.21073
JUND    1.568763 7.774542 12.66888 1.639179e-14 5.576896e-11 23.04095
ARHGDIB 1.829897 8.873965 12.09001 6.146497e-14 1.672954e-10 21.75189
LST1    2.556890 6.774452 11.68693 1.580100e-13 3.583929e-10 20.78367</code></pre>
</div>
</div>
<p>The variables included are:</p>
<ul>
<li><code>logFC</code>: log fold change. The sign is relative to your reference. For example, negative logFC for mtbMtb means Mtb minus Media is negative and thus, expression is lower in Mtb-infected samples.</li>
<li><code>AveExpr</code>: average expression across all samples</li>
<li><code>t</code>: test statistic for significance</li>
<li><code>P.Value</code></li>
<li><code>adj.P.Val</code>: FDR adjusted P-value</li>
<li><code>B</code>: beta coefficient</li>
</ul>
<p>With some <code>tidyverse</code> manipulation, we can get results for all genes and variables in one table. Or you can use the <code>kimma</code> function <code>extract_lmFit</code> and skip the coding! This also renames the columns to match <code>kimma</code>’s output for easier model comparison later on.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>fdr_limma <span class="ot">&lt;-</span> <span class="fu">extract_lmFit</span>(<span class="at">design =</span> mm_limma, <span class="at">fit =</span> efit_limma)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(fdr_limma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "lm"     "lm.fit"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fdr_limma<span class="sc">$</span>lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  model    gene variable estimate         pval          FDR        t        B
1 limma  PLXDC2   mtbMtb 4.192477 1.083979e-15 1.218267e-11 13.91593 25.45526
2 limma   PLCB2   mtbMtb 3.323085 1.790385e-15 1.218267e-11 13.67945 24.99951
3 limma CLEC12A   mtbMtb 3.679747 9.731346e-15 4.414463e-11 12.90214 23.21073
4 limma    JUND   mtbMtb 1.568763 1.639179e-14 5.576896e-11 12.66888 23.04095
5 limma ARHGDIB   mtbMtb 1.829897 6.146497e-14 1.672954e-10 12.09001 21.75189
6 limma    LST1   mtbMtb 2.556890 1.580100e-13 3.583929e-10 11.68693 20.78367
   AveExpr
1 6.786156
2 6.618024
3 5.901780
4 7.774542
5 8.873965
6 6.774452</code></pre>
</div>
</div>
</section>
<section id="paired-sample-design-in-limma" class="level3">
<h3 class="anchored" data-anchor-id="paired-sample-design-in-limma">Paired sample design in <code>limma</code></h3>
<p><code>limma</code> uses a shortcut to model paired sample designs. Unlike a true mixed effect model, <code>limma</code> estimates the mean correlation of gene expression between pairs. This is an approximation of a mixed effects model. While it runs very fast, it assumes the paired design impacts all genes equally</p>
<p>Using the same model as before, we can calculate the mean correlation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>consensus.corr <span class="ot">&lt;-</span> <span class="fu">duplicateCorrelation</span>(<span class="at">object =</span> dat_tcell<span class="sc">$</span>E, </span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">design =</span> mm_limma,</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">block =</span> dat_tcell<span class="sc">$</span>targets<span class="sc">$</span>ptID)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>consensus.corr<span class="sc">$</span>consensus.correlation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3554838</code></pre>
</div>
</div>
<p>You can then incorporate this estimate into the <code>limma</code> model. We will not do not here, because we now have full mixed effects models in <code>kimma</code>.</p>
</section>
</section>
<section id="modeling-in-kimma" class="level2">
<h2 class="anchored" data-anchor-id="modeling-in-kimma">Modeling in <code>kimma</code></h2>
<p><code>kimma</code> supports more flexible modeling of RNA-seq data including simple linear and linear mixed effects models with co-variates, weights, random effects, and covariance matrices. Let’s run the same models as we did with <code>limma</code>.</p>
<p><em>Note that <code>kimma</code> is slower than <code>limma</code>, because it runs a true mixed effects model as well as can run multiple models at once. It can be run on multiple processors to increase speed. </em></p>
<p>Here, we stick to 4 processor to not overload the server. If you’re running this locally, you can omit the <code>processors</code> option and <code>kimma</code> will automatically run on all processors minus 2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>fit_kimma <span class="ot">&lt;-</span> <span class="fu">kmFit</span>(<span class="at">dat =</span> dat_tcell, </span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">model =</span> <span class="st">"~ mtb + (1|ptID)"</span>, </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">use_weights =</span> <span class="cn">TRUE</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">run_lm =</span> <span class="cn">TRUE</span>, <span class="at">run_lme =</span> <span class="cn">TRUE</span>,</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">metrics =</span> <span class="cn">TRUE</span>,</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">processors =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>lm model: expression~mtb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>lme/lmerel model: expression~mtb+(1|ptID)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Input: 30 libraries from 15 unique patients</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Model: 30 libraries</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Complete: 13609 genes</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Failed: 0 genes</code></pre>
</div>
</div>
<p>The <code>kimma</code> output contains 4 data frames: one for each model’s results (like <code>limma</code>’s <code>topTable</code>) and one for each model’s fit metrics, which unlike <code>limma</code>, contains several fit metrics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(fit_kimma)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "lm"      "lme"     "lm.fit"  "lme.fit"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fit_kimma<span class="sc">$</span>lm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 8
  model gene    variable       df statistic estimate    pval     FDR
  &lt;chr&gt; &lt;chr&gt;   &lt;chr&gt;       &lt;int&gt;     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;
1 lm    A1BG    (Intercept)    28    -2.13    -0.491 0.0422  0.0529 
2 lm    A1BG    mtbMtb         28    -1.35    -0.430 0.188   0.312  
3 lm    A2ML1   (Intercept)    28    -1.49    -0.451 0.147   0.169  
4 lm    A2ML1   mtbMtb         28    -0.960   -0.401 0.345   0.485  
5 lm    A3GALT2 (Intercept)    28    -3.22    -0.738 0.00321 0.00462
6 lm    A3GALT2 mtbMtb         28    -0.706   -0.223 0.486   0.618  </code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(fit_kimma<span class="sc">$</span>lm.fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  model  gene    sigma   AIC   BIC     Rsq  adj_Rsq
  &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;
1 lm.fit A1BG    0.856  83.3  87.5 0.0612   0.0277 
2 lm.fit A2ML1   1.12   99.5 104.  0.0319  -0.00268
3 lm.fit A3GALT2 0.853  83.0  87.2 0.0175  -0.0176 
4 lm.fit A4GALT  2.18   87.6  91.8 0.109    0.0771 
5 lm.fit AAAS    0.640  35.1  39.4 0.00395 -0.0316 
6 lm.fit AACS    0.853  49.1  53.3 0.00262 -0.0330 </code></pre>
</div>
</div>
</section>
<section id="picking-a-best-fit-model" class="level2">
<h2 class="anchored" data-anchor-id="picking-a-best-fit-model">Picking a best fit model</h2>
<p>We can now look at metrics like AIC where we see that best fit varies by gene (which is very common)…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>fit_kimma_all <span class="ot">&lt;-</span> <span class="fu">full_join</span>(fit_kimma<span class="sc">$</span>lm.fit, </span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>                           fit_kimma<span class="sc">$</span>lme.fit, </span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"gene"</span>), </span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">"_lm"</span>,<span class="st">"_lme"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#create color variable</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">diff =</span> AIC_lme<span class="sc">-</span>AIC_lm,</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">diff_col =</span> <span class="fu">case_when</span>(diff<span class="sc">&lt;=-</span><span class="dv">7</span> <span class="sc">|</span> diff<span class="sc">&gt;=</span><span class="dv">7</span> <span class="sc">~</span> <span class="st">"Strong"</span>,</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>                              diff<span class="sc">&lt;=-</span><span class="dv">2</span> <span class="sc">|</span> diff<span class="sc">&gt;=</span><span class="dv">2</span> <span class="sc">~</span> <span class="st">"Moderate"</span>,</span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>                              <span class="cn">TRUE</span><span class="sc">~</span><span class="st">"No difference"</span>),</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">diff_col =</span> <span class="fu">factor</span>(diff_col, </span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a>                           <span class="at">levels=</span><span class="fu">c</span>(<span class="st">"No difference"</span>,</span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a>                                    <span class="st">"Moderate"</span>,<span class="st">"Strong"</span>)))</span>
<span id="cb46-13"><a href="#cb46-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-14"><a href="#cb46-14" aria-hidden="true" tabindex="-1"></a>fit_kimma_all <span class="sc">%&gt;%</span></span>
<span id="cb46-15"><a href="#cb46-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> AIC_lm, <span class="at">y =</span> AIC_lme)) <span class="sc">+</span></span>
<span id="cb46-16"><a href="#cb46-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="fu">aes</span>(<span class="at">color=</span>diff_col)) <span class="sc">+</span></span>
<span id="cb46-17"><a href="#cb46-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb46-18"><a href="#cb46-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb46-19"><a href="#cb46-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>() <span class="sc">+</span></span>
<span id="cb46-20"><a href="#cb46-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"AIC"</span>, <span class="at">color=</span><span class="st">"AIC difference"</span>) <span class="sc">+</span></span>
<span id="cb46-21"><a href="#cb46-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"grey40"</span>,<span class="st">"orange"</span>,<span class="st">"darkred"</span>)) <span class="sc">+</span></span>
<span id="cb46-22"><a href="#cb46-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x=</span><span class="dv">150</span>, <span class="at">y=</span><span class="dv">0</span>, <span class="at">label=</span><span class="st">"Better fit by lme"</span>)<span class="sc">+</span></span>
<span id="cb46-23"><a href="#cb46-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x=</span><span class="dv">0</span>, <span class="at">y=</span><span class="dv">150</span>, <span class="at">label=</span><span class="st">"Better fit by lm"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="2.Bulk_RNAseq_deg_analysis_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>and the overall AIC mean and total are somewhat lower for the simple linear model without paired design.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Mean</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(fit_kimma<span class="sc">$</span>lm.fit<span class="sc">$</span>AIC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 65.0011</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(fit_kimma<span class="sc">$</span>lme.fit<span class="sc">$</span>AIC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 67.92518</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Sum</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(fit_kimma<span class="sc">$</span>lm.fit<span class="sc">$</span>AIC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 884600</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(fit_kimma<span class="sc">$</span>lme.fit<span class="sc">$</span>AIC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 924393.8</code></pre>
</div>
</div>
<p>In general, differences in mean AIC &lt; 2 show that either model is appropriate, differences from 2 to 7 are moderate evidence for the lower AIC model, and differences greater than 7 are strong evidence for the lower AIC model.</p>
<p>So in this case, which model do we go with? AIC slightly supports the simple model but our study design is paired… Always use your scientific reasoning first! If you know that there is a study design feature or confounding covariate, keep them in the model even if the AIC says otherwise. Be smarter than your data!</p>
<p>For this experiment, we know we have a paired design so either <code>limma</code> with <code>duplicateCorrelation</code> or <code>kimma</code> with <code>run.lme</code> is appropriate. In our experience, a true mixed effects model in <code>kimma</code> yields fewer false positive genes when you have a paired design, even if metrics like AIC do not support it as the best fit model.</p>
</section>
<section id="significant-genes" class="level2">
<h2 class="anchored" data-anchor-id="significant-genes">Significant genes</h2>
<p>Here, we summarize how many genes are significant at different FDR cutoffs. We see a strong Mtb effect.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summarise_kmFit</span>(fit_kimma<span class="sc">$</span>lme)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 7
  variable          `fdr&lt;0.05` `fdr&lt;0.1` `fdr&lt;0.2` `fdr&lt;0.3` `fdr&lt;0.4` `fdr&lt;0.5`
  &lt;fct&gt;                  &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;
1 (1 | ptID)               958      2117      3526      4675      5667      6548
2 mtb                     6884      7759      8777      9589     10186     10827
3 total (nonredund…       7242      8485      9822     10782     11417     12011</code></pre>
</div>
</div>
<p>Because there are so many significant genes, it can be difficult to interpret results. You’ll see further methods this afternoon!</p>
</section>
<section id="navigation" class="level1">
<h1><em>Navigation</em></h1>
<ul>
<li>Previous lesson: <a href="https://fredhutch.github.io/seatrac-hackday-2023/1.rnaseq_tutorial/1.setup.html">Setup instructions</a></li>
<li>Next lesson: <a href="https://fredhutch.github.io/seatrac-hackday-2023/1.rnaseq_tutorial/3.scRNAseq_analysis.html">Single cell RNAseq</a></li>
<li><a href="https://github.com/FredHutch/seatrac-hackday-2023">Hackday GitHub</a></li>
</ul>
</section>
<section id="additional-resources" class="level1">
<h1><em>Additional resources</em></h1>
<ul>
<li><a href="https://bigslu.github.io/workshops/">More workshops</a></li>
<li><a href="https://bigslu.github.io/tutorials/">More tutorials</a></li>
<li><a href="http://bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma manual, Chapter 15</a></li>
<li><a href="https://bigslu.github.io/kimma_vignette/">kimma vignette</a></li>
</ul>
<hr>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>